feat: Add Brew module

mecattaf opened the pull request:
## Summary



This pull request adds a new Brew module to the BlueBuild project. The Brew module installs Homebrew (Brew) on the system, installs specified Brew packages, and configures systemd services to maintain the packages.



## Features



- Installs Brew at build time.

- Configures and installs specified Brew packages.

- Sets up systemd services to update and upgrade Brew packages automatically.

- Options to control the frequency of updates and upgrades.



## Configuration Options



- `install_brew` (optional: boolean, default: true): Determines whether Brew should be installed at build time.

- `packages` (required: list of strings): A list of Brew packages to be installed.

- `update_interval` (optional: string, default: '6h'): Defines how often the Brew update service should run.

- `upgrade_interval` (optional: string, default: '8h'): Defines how often the Brew upgrade service should run.

- `wait_after_boot_update` (optional: string, default: '10min'): Time delay after system boot before the first Brew update runs.

- `wait_after_boot_upgrade` (optional: string, default: '30min'): Time delay after system boot before the first Brew upgrade runs.

- `disable_update` (optional: boolean, default: false): If set to `true`, disables the automatic activation of the `brew-update.timer`.

- `disable_upgrade` (optional: boolean, default: false): If set to `true`, disables the automatic activation of the `brew-upgrade.timer`.



## Example Configuration



Here is an example configuration for the Brew module in `recipe.yml`:



```yaml

---

type: brew

# Installs Brew (Homebrew) to /home/linuxbrew/.linuxbrew via external download

install_brew: true # Optional - Default: true

# List of Brew packages to be installed

packages:

  - ollama # Required

# Interval between Brew updates

update_interval: '6h'  # Optional - Default: '6h'

# Interval between Brew package upgrades

upgrade_interval: '8h' # Optional - Default: '8h'

# Time delay after boot before first Brew update

wait_after_boot_update: '10min' # Optional - Default: '10min'

# Time delay after boot before first Brew upgrade

wait_after_boot_upgrade: '30min' # Optional - Default: '30min'

# If true, disables automatic activation of `brew-update.timer`.

disable_update: false # Optional - Default: false

# If true, disables automatic activation of `brew-upgrade.timer`.

disable_upgrade: false # Optional - Default: false

```



## Motivation



Homebrew provides a vast collection of CLI applications that might not be available in Fedora repositories. This module leverages Homebrew to provide easy access to these applications.



## Additional Notes



This module follows the recommendations of the Brew maintainers, specifically targeting a single user (UID 1000) to run Brew commands. The module sets up systemd timers to ensure Brew packages remain updated and upgraded over time.



## Context



Context for this module's implementation was based on previous discussions and guidance provided by @m2Giles and other BlueBuild maintainers regarding the installation and management of Brew.



Related Issue: [#230](https://github.com/blue-build/modules/issues/230)



<details>

<summary>Interpretation of Bluefin Brew PR Files</summary>



### Systemd Services and Timers



- **brew-setup.service**:

  - Sets up Brew after installation.



- **brew-update.timer**:

  - Runs periodically to update Brew.



- **brew-upgrade.timer**:

  - Runs periodically to upgrade Brew packages.



- **brew-update.service**:

  - Service unit executed by the `brew-update.timer`.



- **brew-upgrade.service**:

  - Service unit executed by the `brew-upgrade.timer`.



### Helper Scripts



- **brew.sh**:

  - Shell script to download and install Brew.



### Configuration for User Shells



- **brew.fish**:

  - Fish shell configuration for Brew.



- **brew-bash-completion.sh**:

  - Bash shell completion configuration for Brew.



### tmpfiles.d Configuration



- **homebrew.conf**:

  - Configuration file to set up necessary directories for Brew.



### Mounts Configuration



- **var-home-linuxbrew.mount**:

  - Mounts for overlay filesystem to support Brew.



</details>



Looking forward to your feedback!

mecattaf committed:
feat(brew): add new Brew module for Homebrew installation and management

mecattaf committed:
feat(brew): add new Brew module for Homebrew installation and management

gmpinder commented on modules/brew/brew.sh:
```diff
@@ -0,0 +1,219 @@
+
+# Copy mounts configuration file
+
+cp -r "$MODULE_DIRECTORY"/brew/var-home-linuxbrew.mount /usr/lib/systemd/system/var-home-linuxbrew.mount
```
You're copying some files, but I don't see where these files are. Did you mean to add them to this PR?

gmpinder commented on modules/brew/module.yml:
```diff
@@ -0,0 +1,43 @@
+
+  # If true, disables automatic activation of `brew-upgrade.timer`.
+
+  disable_upgrade: false # Optional - Default: false - Expects type: boolean
```
We have a standard to use `kebab-case` for the properties of modules. Mind updating them to be that way?

fiftydinar replied:
This is not solved as of now, so I'm opening the conversation.

mecattaf replied:
module.yml now uses kebab-case: https://github.com/blue-build/modules/pull/239/files#diff-e321f3fe6fca7490fb93078c575852784e86eccb2a746e9f33b29352444112f0

fiftydinar replied:
Please update the README part that is highlighted in this review with kebab-cases

fiftydinar replied:
I fixed it.

gmpinder commented on modules/brew/README.md:
```diff
@@ -0,0 +1,147 @@
+
+
+
+## Features
```
That's a lot of whitespace. Mind cleaning that up?

gmpinder commented on modules/brew/README.md:
```diff
@@ -0,0 +1,147 @@
+
+
+
+A list of Brew packages to be installed. This is a mandatory configuration and must be provided.
```
Should this be a requirement? I think we should allow the user to decide to install something at build time vs runtime.

xynydev replied:
+1

gmpinder reviewed (changes requested):
I really like this idea and will most likely use this for myself. Just had some nit-picks from me. @fiftydinar and @xynydev will probably want to dig deeper into this.

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,147 @@
+
+
+
+### `install_brew` (optional: boolean, default: true)
```
What is the module used for if not installing brew? How do we expect the services to add brew packages and update brew to work if brew isn't installed?

Is there a possibility of conflict if a user has brew installed prior to switching to an image with brew pre-installed, which would make not pre-installing brew while using the module a valid choice? If so, we should document that very clearly as a warning.

IMO (and with the limited information I have), just including `- type: brew` should be enough to install brew, such as including `- type: yafti` installs yafti.

mecattaf replied:
> What is the module used for if not installing brew? How do we expect the services to add brew packages and update brew to work if brew isn't installed?
> 
> Is there a possibility of conflict if a user has brew installed prior to switching to an image with brew pre-installed, which would make not pre-installing brew while using the module a valid choice? If so, we should document that very clearly as a warning.
> 
I drew inspiration from the [chezmoi module](https://github.com/blue-build/modules/blob/main/modules/chezmoi/module.yml#L7), which does have an "install" boolean. 

> IMO (and with the limited information I have), just including `- type: brew` should be enough to install brew, such as including `- type: yafti` installs yafti.

I agree, will remove that option :+1:

xynydev replied:
> I drew inspiration from the chezmoi module, which does have an "install" boolean.

That module has some issues too, which unfortunately made it through review. Trying to catch those before merge in the future.

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,147 @@
+
+
+
+### `disable_update` (optional: boolean, default: false)
```
This and `disable_upgrade` could be `auto-update` and `auto-upgrade`, which would be `true` by default and could be set to `false` by the user. That would be more clear IMO.

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,147 @@
+
+
+
+## Example Configuration
```
The example configuration is automatically added to the website documentation / reference page from `module.yml` and shouldn't be added to the README

m2Giles commented:
The var-home-linuxbrew.mount is vestigial. Recommend using tmpfiles.d instead for creating the copy to point.

Ultimately on first boot you want the contents of /usr/share/homebrew to end up at /var/home/linuxbrew.

If you go the mount route, you will need an overlayfs since /usr is read only. Eventually everything will be pulled up to the upper dir so you are not really saving space. It's also being stored in /usr compressed as well.

I'm also unsure why you would have a boolean for installing or not. Should be if they want this or not. Every service file is dependent on brew being present.

xynydev commented:
> The var-home-linuxbrew.mount is vestigial. Recommend using tmpfiles.d instead for creating the copy to point.
> 
> Ultimately on first boot you want the contents of /usr/share/homebrew to end up at /var/home/linuxbrew.
> 
> If you go the mount route, you will need an overlayfs since /usr is read only. Eventually everything will be pulled up to the upper dir so you are not really saving space. It's also being stored in /usr compressed as well.

+1, I had the exact same thoughts, but didn't know how to comment yet, without seeing the service units that haven't been pushed to the PR yet. 

Has this module been tested yet? I don't expect `brew install` to work with the current way things work.

fiftydinar commented:
Thanks for the contribution!

I currently don't have anything to add, since xynydev & gmpinder said everything that I see as flaws that have to be fixed.

m2giles also added a good point about using mount vs tmpfiles.d.

mecattaf committed:
feat(brew): part 2 of adding new Brew module for Homebrew installation and management

mecattaf commented:
Thanks for all the feedback. I think the updates I just pushed should cover everything.

mecattaf committed:
feat(brew): part 2.2 of adding new Brew module for Homebrew installation and management

mecattaf reviewed (commented):
Implemented the changes you requested

gerblesh commented on modules/brew/brew.fish:
would probably rename this to `brew-fish-completions.fish` for consistency sake, and to not confuse with `brew.sh`

gerblesh commented on modules/brew/README.md:
```diff
@@ -0,0 +1,36 @@
+If true, disables automatic activation of `brew-upgrade.timer`.
+
+## Development
+
```
sorta a nitpick again, but the whitespace needs fixing here

gerblesh reviewed (changes requested):
Can't wait for this to get merged! Just had some minor nitpicks, but overall looks like a great module

m2Giles commented on modules/brew/brew-setup.service:
```diff
@@ -0,0 +1,18 @@
+Type=oneshot
+ExecStart=/usr/bin/mkdir -p /tmp/homebrew
+ExecStart=/usr/bin/tar --zstd -xvf /usr/share/homebrew.tar.zst -C /tmp/homebrew
+ExecStart=/usr/bin/cp -R -n /tmp/homebrew/home/linuxbrew/.linuxbrew /var/home/linuxbrew
```
/home/linuxbrew doesn't exist by default. You need to make it prior to this copy. You can either do a mkdir here or use tmpfiles.d

fiftydinar replied:
It looks like it's being in `tmpfiles.d` already?

https://github.com/blue-build/modules/pull/239/files#diff-a34ba5d080c32ad5ffa035babc519b97dd8596065fcd1b9a3b48e74b79fc4450R3

fiftydinar commented on modules/brew/README.md:
```diff
@@ -0,0 +1,36 @@
+### `wait_after_boot_upgrade` (optional: string, default: '30min')
+Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).
+
+### `auto_update` (optional: boolean, default: false)
```
`auto-update` & `auto-upgrade` should be updated to contain `default: true` in README

fiftydinar commented on modules/brew/README.md:
```diff
@@ -0,0 +1,36 @@
+## Features
+- Installs Brew at build time.
+- Configures and installs specified Brew packages.
+- Sets up systemd services to update and upgrade Brew packages automatically.
```
It should be more clearly noted about what is the difference between `update` & `upgrade` in README.

Most `update` & `upgrade` commands are the same with aliases, but that is not the case here.

`update` updates the Brew binary to latest version, while `upgrade` updates the installed Brew packages to latest version.

fiftydinar commented on modules/brew/brew-upgrade.service:
```diff
@@ -0,0 +1,10 @@
+[Service]
+# Override the user if different UID/User
+User=1000
+Type=ones
```
Is this a typo to `oneshot`?

fiftydinar replied:
Yeah, looks like it's not properly copied from Bluefin upgrade service.

```
[Unit]
Description=Upgrade Brew packages
After=local-fs.target
After=network-online.target
ConditionPathIsSymbolicLink=/home/linuxbrew/.linuxbrew/bin/brew

[Service]
# Override the user if different UID/User
User=1000
Type=oneshot
Environment=HOMEBREW_CELLAR=/home/linuxbrew/.linuxbrew/Cellar
Environment=HOMEBREW_PREFIX=/home/linuxbrew/.linuxbrew
Environment=HOMEBREW_REPOSITORY=/home/linuxbrew/.linuxbrew/Homebrew
ExecStart=/usr/bin/bash -c "/home/linuxbrew/.linuxbrew/bin/brew upgrade"
```

fiftydinar commented on modules/brew/module.yml:
```diff
@@ -0,0 +1,22 @@
+shortdesc: The brew module installs Homebrew (Brew) at build time and ensures the package manager remains updated and maintained.
+
+readme: https://raw.githubusercontent.com/blue-build/modules/main/modules/brew/README.md
+
```
Whitespaces need to be solved here.

fiftydinar replied:
Solved.

mecattaf committed:
feat(brew): part 3 of adding new Brew module for Homebrew installation and management

fiftydinar committed:
docs: Add kebab-cases for module entries in README

fiftydinar committed:
docs: Don't add whitespaces in module.yml

fiftydinar commented on modules/brew/brew.sh:
```diff
@@ -0,0 +1,101 @@
+
+# Copy systemd service and timer files
+cp -r "$MODULE_DIRECTORY"/brew/brew-setup.service /usr/lib/systemd/system/brew-setup.service
+cp -r "$MODULE_DIRECTORY"/brew/brew-update.service /usr/lib/systemd/system/brew-update.service
```
Logic for writing custom time values from `UPDATE_INTERVAL`, `UPGRADE_INTERVAL`, `$WAIT_AFTER_BOOT_UPDATE` & `$WAIT_AFTER_BOOT_UPGRADE` to systemd units is missing.

mecattaf replied:
Nice catch! 
I think the best way to approach this is to write the services from the brew.sh script instead of copying over the files themselves. Lmk if you think this is the right way to proceed. 

As seen in the [chezmoi module](https://github.com/blue-build/modules/blob/main/modules/chezmoi/chezmoi.sh)
```
if [[ $DISABLE_INIT == false ]]; then
	# Write the service to initialize Chezmoi, and insert the repo url in the file
	echo "Writing init service to user unit directory"
	cat >>/usr/lib/systemd/user/chezmoi-init.service <<EOF
  [Unit]
  Description=Initializes Chezmoi if directory is missing
  
  # This service will not execute for a user with an existing chezmoi directory
  ConditionPathExists=!%h/.local/share/chezmoi/.git/
  [Service]
  ExecStart=/usr/bin/chezmoi init --apply ${DOTFILE_REPOSITORY}
  Type=oneshot

  [Install]
  WantedBy=default.target
EOF
```

fiftydinar replied:
I agree with this approach.

mecattaf replied:
For consistency the brew.sh script is now responsible for writing the following files:
- `brew-update.timer `
- `brew-upgrade.timer `
- `brew-setup.service `
- `brew-update.service `
- `brew-upgrade.service`

We chose to dynamically generate all systemd units, even if some may not be directly affected by custom flags ("install service" nor "upgrade service"). Those files were removed from the module, which makes it easier to understand.

fiftydinar replied:
@mecattaf
Looks good, except files are not physically removed. You just removed lines where those files are copied.

Just do this & it's solved.

fiftydinar commented on modules/brew/brew.sh:
```diff
@@ -0,0 +1,101 @@
+# Install specified Brew packages if any
+if [[ -n $PACKAGE_LIST ]]; then
+    echo "Installing specified Brew packages..."
+    su -c "/home/linuxbrew/.linuxbrew/bin/brew install $PACKAGE_LIST" -s /bin/bash linuxbrew
```
While this might work, I believe that using array is more reliable here.

I can do changes on this if you're not sure how to handle this.

mecattaf replied:
Please do

fiftydinar replied:
Did it.

fiftydinar committed:
chore: Use arrays for package list & add quotes + brackets to some strings

to make it more safe.

mecattaf committed:
feat(brew): systemd services and timers written dynamically

mecattaf committed:
feat(brew): systemd services and timers written dynamically 2

fiftydinar commented on modules/brew/brew.sh:
```diff
@@ -0,0 +1,177 @@
+EOF
+
+# Copy shell configuration files
+cp -r "${MODULE_DIRECTORY}"/brew/brew-fish-completions.fish /usr/share/fish/vendor_conf.d/brew-fish-completions.fish
```
Nitpick, but please rename this line & the file to `brew-fish-completion.fish` instead of `brew-fish-completions.fish`

Bash completion is named `brew-bash-completion.sh`, so it makes sense to rename it appropriately.

xynydev replied:
Isn't `conversations` the proper form? I feel like I've seen that more often.

fiftydinar replied:
@xynydev

`completions` might make more sense when I think about it, since it indicates various command completions for `brew`. Command completion indicates that brew can only execute 1 command which gets the completion, which is not true.

Easy change, I'll do it.

fiftydinar replied:
Did it

mecattaf committed:
feat(brew): clean up fish completion file name

fiftydinar committed:
chore: Remove duplicate brew completions file

fiftydinar commented:
After all of this fixing, this looks good to me!

The only thing that remains is to test this in real-life, to check if everything is working correctly.

fiftydinar commented:
Brew install fails, because execution script detects that it's ran as root:

```
0.091 ================================================= Start 'brew' Module =================================================
0.126 Downloading and installing Brew...
0.153 Warning: Running in non-interactive mode because `stdin` is not a TTY.
0.174 ==> Checking for `sudo` access (which may request your password)...
0.175 Don't run this as root!
```

I tested build with `type: brew` added.

Build has this PR commit as the latest:
[`91eaad0` (#239)](https://github.com/blue-build/modules/pull/239/commits/91eaad0e33fc9bf3e2a94b789fa1853c5dbdc364)

Failed build:
https://github.com/fiftydinar/gidro-os/actions/runs/9272606213/job/25510790662#step:2:2331

fiftydinar commented:
I think these ones are responsible to trick `brew` that build is not running as root:

https://github.com/ublue-os/bluefin/blob/main/build_files/base/brew.sh#L5C1-L6C18

https://github.com/ublue-os/bluefin/blob/main/scripts/sudoif.sh

mecattaf committed:
feat(brew): tweak in brew.sh 3

fiftydinar committed:
Merge branch 'main' into add-brew-module

fiftydinar committed:
chore: Rename shell completion to completions pt. 1

fiftydinar committed:
chore: Rename shell completion to completions pt. 2

fiftydinar committed:
chore: Rename shell completion to completions pt. 3

mecattaf commented:
In brew.sh 
- added `touch` like bluefin
- renamed brew.sh to brew-install inside the script to avoid confusion
- made the homebrew.conf file dynamically written, like [this](https://github.com/rsturla/eternal-images/blob/b1a44251d203c52f5dd0bb2f2568195159bc94b9/lumina/scripts/_base/012-install-brew.sh#L29)

Mind trying now?

fiftydinar commented:
@mecattaf
Installation passes successfully now in logs (`type: brew` used only).

Logs are very noisy, but maybe necessary?
https://github.com/fiftydinar/gidro-os/actions/runs/9273300849/job/25513073884?pr=26#step:2:2146

I will test the image in a VM.

fiftydinar commented on modules/brew/README.md:
```diff
@@ -0,0 +1,36 @@
+# brew
```
# Nofile limits

I think that we should document the necessary tweak to install/run some brew packages. Else, some packages complain about file limits & fail to run.

It is necessary to increase nofile limits, like it's done here (both limits.d & systemd config should be modified):


- SSH/tty nofile limit
`/usr/etc/security/limits.d/30-brew-limits.conf`

```
#This file sets the resource limits for the users logged in via PAM,
#more specifically, users logged in on via SSH or tty (console).
#Limits related to terminals in Wayland/Xorg sessions depend on a
#change to /etc/systemd/user.conf.
#This does not affect resource limits of the system services.
#This file overrides defaults set in /etc/security/limits.conf

* soft nofile 4096
* hard nofile 524288
```

- SystemD nofile limit affecting whole session
`/usr/lib/systemd/system/system.conf.d/30-brew-limits.conf`
`/usr/lib/systemd/user/user.conf.d/30-brew-limits.conf`

```
[Manager]
DefaultLimitNOFILE=4096:524288
```

# Necessary dependencies

Document that GCC needs to be installed in the image for brew to work.
I read that brew can provide gcc too, but I would let OS package manager to handle this one, since GCC is available in basically all distros.

mecattaf replied:
Should we add those overrides as lines of `cat` in brew.sh, the same way we did for the systemd units? 


> # Necessary dependencies
> Document that GCC needs to be installed in the image for brew to work. I read that brew can provide gcc too, but I would let OS package manager to handle this one, since GCC is available in basically all distros.

Recommended way: 
```
rpm:
  install:
    - gcc
```

fiftydinar replied:
I'm not sure if we should automatically add those for all users, since increasing nofile limits can have some potential security implications if application chooses to abuse it for evil purposes. I'm not the security expert, so can't explain it in detail, but that's how I see it.

It can also potentially override the same modification that someone did differently in their image for some other purposes.

I think that just recommendation in docs is better.

About GCC, I think that we can check if `gcc` command exists. If it doesn't, than inform the user that `gcc` is not installed, how it must be installed for brew to function & fail the build. We wouldn't add the else statement to do `rpm-ostree install`, since we try to make module to be universally compatible regarding container-based OSes (like Fedora Atomic & VanillaOS in the future)
Of course, we would still document this `gcc` requirement in docs.

mecattaf replied:
I think these two points can be resolved by improving brew.sh.
We could a flag that manages nofile-limits? It would default to False, and clearly mention the potential security implications if True.

fiftydinar replied:
I'll let others also discuss this review discussion & than try together to find the sweet-spot solution.

mecattaf committed:
feat(brew): Nofile Limits flag, update brew.sh module.yml and README.md

mecattaf committed:
feat(brew): Brew analytics opt-out flag

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,45 @@
+Determines whether to apply nofile limits for Brew installations. When set to true, it increases the nofile limits to prevent certain packages from failing due to file limits. However, it's important to note that increasing nofile limits can have potential security implications. Defaults to false for security purposes.
+
+### `brew-analytics` (optional: boolean, default: true)
+Determines whether to opt-out of Brew analytics. When set to true, analytics are disabled.
```
"When set to *true* analytics are *disabled*."

What? Mistake in the docs perhaps?

mecattaf committed:
Brew Analytics Enabled by default in docs

fiftydinar commented on modules/brew/brew.sh:
```diff
@@ -212,6 +217,12 @@ DefaultLimitNOFILE=4096:524288
+# Disable homebrew analytics if the flag is set to false
+# like secureblue: https://github.com/secureblue/secureblue/blob/live/config/scripts/homebrewanalyticsoptout.sh
+if [[ "${BREW_ANALYTICS}" == false ]]; then
+    echo "HOMEBREW_NO_ANALYTICS=1" >> /usr/etc/environment
```
I think there could be a better way to set environment variable to be only available when in terminal session in `profiles.d`. Similarly to how is brew path preferred only when in terminal session.

Making it available globally in `/usr/etc/environment` is not the ideal solution imo.

fiftydinar replied:
Hmm, a better solution would be to maybe alias `brew` to always launch with `HOMEBREW_NO_ANALYTICS` set up.

Having it applied in terminal session only is flawed, since I forgot that brew runs in background to update & upgrade.

fiftydinar replied:
If we could fix this error:

`#46 14.06 su: user linuxbrew does not exist or the user entry does not contain all the required fields`

I think that we could just execute

`su -c "/home/linuxbrew/.linuxbrew/bin/brew analytics off" -s /bin/bash linuxbrew`

to turn off analytics. This would be the best way, since it does not touch environment & shell, but the app itself.

fiftydinar replied:
I ended up just improving the script a bit, to account for existing homebrew_analytics variable, so it doesn't get duplicated.

I think that there is not a better solution for doing this, so I'll mark this as resolved.

fiftydinar commented on modules/brew/module.yml:
```diff
@@ -0,0 +1,23 @@
+example: |
+  type: brew
+  # List of Brew packages to be installed
+  packages:
```
I tested the image.

When using `type: brew`, it works nicely, as it should.

`brew doctor` reports no errors & packages install successfully.

However, when I try to install packages in the image inside `packages:` array, it doesn't work and fails with this error:
`#46 14.06 su: user linuxbrew does not exist or the user entry does not contain all the required fields`

Build log:
https://github.com/fiftydinar/gidro-os/actions/runs/9274948443/job/25518506501?pr=26#step:2:9056

fiftydinar replied:
@mecattaf Apologize, I removed the comment & migrated it here.

I'll quote you:

> Maybe we should not install brew packages during the build stage. This module can behave like the [flatpak module](https://github.com/blue-build/modules/blob/main/modules/default-flatpaks/user-flatpak-setup?rgh-link-date=2024-05-28T19%3A31%3A03Z): install and update the brew packages listed in recipe.yml while the device is running.

I don't know how it works for Bluefin, I would be interested to investigate.

fiftydinar replied:
Just checked it, Bluefin doesn't install brew packages in build-time yet.

I think this happens because user is technically not created, as doing mkdir to `/var/home/linuxbrew/` & setting some permissions is not enough.

I'm not sure if we should implement installing packages when system is booted, effectively making it a boot-time module.

I'll think about how to solve this in build-time, but I'm not sure if it's possible.

fiftydinar replied:
I disabled installing brew packages in build-time until this is figured out.

fiftydinar replied:
We won't implement installing brew packages for now.

It can be always added in the future if possible & if desired.

fiftydinar commented on modules/brew/brew.sh:
```diff
@@ -11,6 +11,13 @@ if [[ "${DEBUG}" == true ]]; then
+# Check if gcc is installed
+if ! command -v gcc &> /dev/null
+then
+    echo "Error: gcc could not be found. Include \"gcc\" in the list of packages to install with rpm-ostree"
```
```suggestion
    echo "ERROR: \"gcc\" package could not be found."
    echo "       Brew depends on \"gcc\" in order to function"
    echo "       Please include \"gcc\" in the list of packages to install with the system package manager"
```

mecattaf commented:
I will defer to more experienced contributors to keep improving the PR, but as I understand TODO:
- [ ] Remove verbose logging from brew.sh (not possible to fix, not a blocker)
- [ ] Resolve installing brew packages at build time (not possible to fix, not a blocker)
- [x] Decide approach re: nofile limit and brew analytics

Thanks

fiftydinar commented on modules/brew/brew.sh:
```diff
@@ -0,0 +1,234 @@
+fi
+
+# Apply nofile limits if enabled
+if [[ "${NOFILE_LIMITS}" == true ]]; then
```
I have an idea on how to improve this process.

We can check the current nofile limit in the base image with:

SSH/tty (not sure if it shows SystemD limit even when not set in limits.d, needs to be tested):
`ulimit -Sn` (soft limit)
`ulimit -Hn` (hard limit)

SystemD:
`systemctl show -all`
(`DefaultLimitNOFILE` = hard limit, `DefaultLimitNOFILESoft` = soft limit)

If the base image has limit less than desired, than do the change.
If it's equal or has more than enough, than do nothing & inform user that no change is needed.

fiftydinar replied:
This is what I came up with.

It's a 231 lines of code, so `source` would be used below `nofile-limits` if statement.

It dynamically searches for CURRENT soft & hard limit between these files & directories, sorted by preference (from least preferred to most preferred):

SSH/TTY limits:
```
"/usr/etc/security/limits.conf"
"/usr/etc/security/limits.d/"
"/etc/security/limits.conf"
"/etc/security/limits.d/"
```

SystemD limits:
```
# System
"/usr/lib/systemd/system.conf"
"/usr/lib/systemd/system.conf.d/"
"/usr/etc/systemd/system.conf"
"/usr/etc/systemd/system.conf.d/"
"/etc/systemd/system.conf"
"/etc/systemd/system.conf.d/"

# User
"/usr/lib/systemd/user.conf"
"/usr/lib/systemd/user.conf.d/"
"/usr/etc/systemd/user.conf"
"/usr/etc/systemd/user.conf.d/"
"/etc/systemd/user.conf"
"/etc/systemd/user.conf.d/"
```

It always shows the CURRENT most preferred soft & hard nofile value.

OPTIMAL nofile limit values are set on these conditions:
- If CURRENT value is equal or more than OPTIMAL, than nothing is done.
- If CURRENT value is less than OPTIMAL, changes are made
- It accounts for the scenario when soft value is OPTIMAL, while hard value is not & vice-versa (so only 1st value is necessarily overwritten, while 2nd value remains unchanged)

File is put in `/usr/lib/systemd/{system,user}.conf.d/`, while `/usr/etc/security/limits.d/` must be used, since `/usr/lib` directory for this doesn't exist.

File is named `zz1-brew-limits.conf`. The reason why it has `zz1-` prefix is to assure that the file will be in highest priority. This takes advantage of lexic sorting, which is preferred in terms of priorities.
While systemd & others recommend using numbers as prefix, we can't guarantee that some image-maintainer doesn't have `99-limits.conf` file, which would override our file.
But most image-maintainers don't know about lexic sorting & using letters as prefixes being possible, so I guess that no image-maintainer at the moment has `zz1-` prefix or higher.

The potential issue can come where image-maintainer applies the modification to `/usr/etc` for systemd modifications. If he modified nofile limits, than his configuration will have higher priority.
However, not a lot of users do this I believe, so just a nitpick.

What do y'all think?

Here's the script:

```bash
#!/usr/bin/env bash

set -euo pipefail

# Script used to read nofile limits from the base image, since issuing easy systemctl status commands don't work in build-time.
# Takes into account config priorities & current config settings regarding nofile limits
# (if nofile limit is already applied with appropriate values in the base image, than this tweak is not applied)
# Modifies limits nofile value & systemd DefaultLimitNOFILE value

DESIRED_SOFT_LIMIT=4096
DESIRED_HARD_LIMIT=524288

# SSH/TTY nofile limit (security ulimit config)

# From least to most preferred
SSH_TTY_LIMIT_ORDER=(
"/usr/etc/security/limits.conf"
"/usr/etc/security/limits.d/"
"/etc/security/limits.conf"
"/etc/security/limits.d/"
)
updated_ssh_array=()
for path in "${SSH_TTY_LIMIT_ORDER[@]}"; do
    if [ -e "${path}" ]; then
        updated_ssh_array+=("${path}")
    fi
done
# Update the original array with the existing paths
SSH_TTY_LIMIT_ORDER=("${updated_ssh_array[@]}")

# Soft SSH/TTY nofile limit
SSH_TTY_SOFT_INFO=$(find "${SSH_TTY_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -v OFS='\t' '/soft\s+nofile/ && !/^#/ {sub(/.*nofile/, ""); gsub(/^[ \t]+/, ""); print FILENAME, $0}' {} + | tail -n 1)
if [[ -n "${SSH_TTY_SOFT_INFO}" ]]; then
  CURRENT_SSH_TTY_SOFT_VALUE=$(echo "${SSH_TTY_SOFT_INFO}" | awk '{print $2}')
else
  CURRENT_SSH_TTY_SOFT_VALUE=0
fi

# Hard SSH/TTY nofile limit
SSH_TTY_HARD_INFO=$(find "${SSH_TTY_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -v OFS='\t' '/hard\s+nofile/ && !/^#/ {sub(/.*nofile/, ""); gsub(/^[ \t]+/, ""); print FILENAME, $0}' {} + | tail -n 1)
if [[ -n "${SSH_TTY_HARD_INFO}" ]]; then
  CURRENT_SSH_TTY_HARD_VALUE=$(echo "${SSH_TTY_HARD_INFO}" | awk '{print $2}')
else
  CURRENT_SSH_TTY_HARD_VALUE=0
fi

# SystemD nofile limit

# SystemD system soft & hard nofile limit
# From least to most preferred
SYSTEMD_SYSTEM_LIMIT_ORDER=(
"/usr/lib/systemd/system.conf"
"/usr/lib/systemd/system.conf.d/"
"/usr/etc/systemd/system.conf"
"/usr/etc/systemd/system.conf.d/"
"/etc/systemd/system.conf"
"/etc/systemd/system.conf.d/"
)
updated_systemd_system_array=()
for path in "${SYSTEMD_SYSTEM_LIMIT_ORDER[@]}"; do
    if [ -e "${path}" ]; then
        updated_systemd_system_array+=("${path}")
    fi
done
# Update the original array with the existing paths
SYSTEMD_SYSTEM_LIMIT_ORDER=("${updated_systemd_system_array[@]}")

# SystemD system soft & hard nofile limit
SYSTEMD_SYSTEM_SOFT_INFO=$(find "${SYSTEMD_SYSTEM_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $2}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_SYSTEM_SOFT_INFO}" ]]; then
  CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE=$(echo "${SYSTEMD_SYSTEM_SOFT_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE=0
fi

SYSTEMD_SYSTEM_HARD_INFO=$(find "${SYSTEMD_SYSTEM_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $3}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_SYSTEM_HARD_INFO}" ]]; then
  CURRENT_SYSTEMD_SYSTEM_HARD_VALUE=$(echo "${SYSTEMD_SYSTEM_HARD_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_SYSTEM_HARD_VALUE=0
fi

# SystemD user soft & hard nofile limit
SYSTEMD_USER_LIMIT_ORDER=(
"/usr/lib/systemd/user.conf"
"/usr/lib/systemd/user.conf.d/"
"/usr/etc/systemd/user.conf"
"/usr/etc/systemd/user.conf.d/"
"/etc/systemd/user.conf"
"/etc/systemd/user.conf.d/"
)
updated_systemd_user_array=()
for path in "${SYSTEMD_USER_LIMIT_ORDER[@]}"; do
    if [ -e "${path}" ]; then
        updated_systemd_user_array+=("${path}")
    fi
done
# Update the original array with the existing paths
SYSTEMD_USER_LIMIT_ORDER=("${updated_systemd_user_array[@]}")

SYSTEMD_USER_SOFT_INFO=$(find "${SYSTEMD_USER_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $2}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_USER_SOFT_INFO}" ]]; then
  CURRENT_SYSTEMD_USER_SOFT_VALUE=$(echo "${SYSTEMD_USER_SOFT_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_USER_SOFT_VALUE=0
fi

SYSTEMD_USER_HARD_INFO=$(find "${SYSTEMD_USER_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $3}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_USER_HARD_INFO}" ]]; then
  CURRENT_SYSTEMD_USER_HARD_VALUE=$(echo "${SYSTEMD_USER_HARD_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_USER_HARD_VALUE=0
fi

# Check current state

echo "Current nofile limit values:"
check_and_print() {
    if [[ "${1}" -eq 0 ]]; then
        echo "UNSET"
    else
        echo "${1}"
    fi
}
echo "SSH/TTY soft nofile limit: $(check_and_print ${CURRENT_SSH_TTY_SOFT_VALUE})"
echo "SSH/TTY hard nofile limit: $(check_and_print ${CURRENT_SSH_TTY_HARD_VALUE})"
echo "SystemD system soft nofile limit: $(check_and_print ${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE})"
echo "SystemD system hard nofile limit: $(check_and_print ${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE})"
echo "SystemD user soft nofile limit: $(check_and_print ${CURRENT_SYSTEMD_USER_SOFT_VALUE})"
echo "SystemD user hard nofile limit: $(check_and_print ${CURRENT_SYSTEMD_USER_HARD_VALUE})"

# Write nofile limit values
# zz1- prefix is used for config, to assure that nofile limit is going to be applied, as it's high in lexical order.
# If config is higher in lexical order, that means that it's preferred over other ones.
# Downstreams can just go on with their own modifications in zz2-, zz3-, zz4-, zz5-, etc. increments
# if they wish to experiment with nofile limits differently

# Write SSH/TTY nolimit values
if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] || [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  if [[ ! -d "/usr/etc/security/limits.d/" ]]; then
    mkdir -p "/usr/etc/security/limits.d/"
  fi
  echo "# This file sets the resource limits for users logged in via PAM,
# more specifically, users logged in via SSH or tty (console).
# Limits related to terminals in Wayland/Xorg sessions depend on a
# change to /etc/systemd/user.conf.
# This does not affect resource limits of the system services.
# This file overrides defaults set in /etc/security/limits.conf

" > "/usr/etc/security/limits.d/zz1-brew-limits.conf"
fi

if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SSH/TTY soft nofile limit value"
  echo "* soft nofile ${DESIRED_SOFT_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]]; then
  echo "Required SSH/TTY soft nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SSH/TTY hard nofile limit value"
  echo "* hard nofile ${DESIRED_HARD_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Required SSH/TTY hard nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SSH/TTY soft & hard nofile limit value"
  echo "* soft nofile ${DESIRED_SOFT_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
  echo "* hard nofile ${DESIRED_HARD_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
fi

# Write SystemD nolimit values

# Writing SystemD system nolimit values
if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] || [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  if [[ ! -d "/usr/lib/systemd/system.conf.d/" ]]; then
    mkdir -p "/usr/lib/systemd/system.conf.d/"
  fi
fi

if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD system soft nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" > "/usr/lib/systemd/system.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]]; then
  echo "Required SystemD system soft nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD system hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/system.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Required SystemD system hard nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD system soft & hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/system.conf.d/zz1-brew-limits.conf"
fi

# Writing SystemD user nolimit values
if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] || [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  if [[ ! -d "/usr/lib/systemd/user.conf.d/" ]]; then
    mkdir -p "/usr/lib/systemd/user.conf.d/"
  fi
fi

if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD user soft nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${CURRENT_SYSTEMD_USER_HARD_VALUE}" > "/usr/lib/systemd/user.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]]; then
  echo "Required SystemD user soft nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD user hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${CURRENT_SYSTEMD_USER_SOFT_VALUE}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/user.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Required SystemD user hard nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD user soft & hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/user.conf.d/zz1-brew-limits.conf"
fi
```

fiftydinar replied:
But, this script is complex as it is, & with potential systemd changes, it could be broken.

So I would maybe lean more towards documenting this, even though I spent some time on this code.

fiftydinar replied:
I merged it in.

I think that this is the best approach for this. It should be future-proof & reliable enough.

fiftydinar commented:
> I will defer to more experienced contributors to keep improving the PR, but as I understand TODO:
> 
>     * [ ]  Remove verbose logging from brew.sh
> 
>     * [ ]  Resolve installing brew packages at build time
> 
>     * [ ]  Decide approach re: nofile limit and brew analytics
> 
> 
> Thanks

That's right.

I think that this can be added in PR post 1st, titled like this:

# To do

- [ ]  Remove verbose logging from brew.sh (not possible to fix, not a blocker)
- [ ]  Resolve installing brew packages at build-time (not possible to fix, not a blocker)
- [x]  Improve approach to solve: nofile limit
- [x]  Improve approach to solve: brew analytics

fiftydinar committed:
docs: `build-time` instead of `build time`

fiftydinar committed:
docs: Improve `gcc` package log explanation

fiftydinar committed:
fix(brew): Fetching YAML entries due to miss of transition from `_` to `-`

fiftydinar committed:
chore(brew): Improve disabling analytics write to `environment`

fiftydinar committed:
chore(brew): Add script for modifying nofile limit

fiftydinar committed:
chore(brew): Forgot to source the nofile script

fiftydinar committed:
chore(brew): Disable option of installing packages pt.1

fiftydinar committed:
chore(brew): Disable option of installing packages pt.2

fiftydinar committed:
chore(brew): Disable option of installing packages pt.3

fiftydinar commented on modules/brew/brew.sh:
```diff
@@ -0,0 +1,216 @@
+
+# Always install Brew
+echo "Downloading and installing Brew..."
+curl -Lo /tmp/brew-install https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh
```
About verbose logging,

I analyzed the install script & I don't see anything that can solve this.

Might as well live with it.

What y'all think?

mecattaf replied:
Great work. 

Once you can confirm that the script works with all the existing flags, @xynydev or @gmpinder can merge this PR.

fiftydinar replied:
Yeah, I will test other things too.

fiftydinar committed:
docs(brew): Remove installing packages as a feature

fiftydinar committed:
chore(brew): Add more logs

fiftydinar committed:
chore(brew): Fix typos in logs

fiftydinar committed:
fix(brew): Use `awk` instead of `grep` which fails for stupid reason

Without any error logs either.

fiftydinar committed:
fix(brew): Update & upgrade permutation

fiftydinar committed:
docs(brew): Fix upgrade & update permutation

fiftydinar committed:
chore(brew): Make small tweaks to brew-analytics

If environment is empty, than don't make the extra whiteline

fiftydinar reviewed (approved):
I tested everything I could & it works!

Maybe @xynydev has something to say about improvement in docs,
but to me, everything looks good!

Approving.

fiftydinar committed:
docs(brew): Slight order adjustment in README inside "Features"

fiftydinar committed:
docs(brew): Add link to Homebrew

fiftydinar commented:
EDIT: Solved

NOTE to add in README:
After excluding `type: brew` from the module after it's installed in the system, some remaining folders will remain in the running system:

# Created by brew
- $HOME/.cache/Homebrew
- and some other

# Created by tmpfiles.d
- /var/lib/homebrew
- /var/cache/homebrew
- /var/home/linuxbrew

This would potentially solve the problem, but it must be executed manually by the user or set as automatic systemd service afterwards by image-maintaner:

```sh
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"
if [[ -d "/var/lib/homebrew" ]]; then
  sudo rm -r /var/lib/homebrew
fi
if [[ -d "/var/cache/homebrew" ]]; then
  sudo rm -r /var/cache/homebrew
fi
if [[ -d "/var/home/linuxbrew" ]]; then
  sudo rm -r /var/home/linuxbrew
fi
```

fiftydinar committed:
docs(brew): Add note about uninstalling brew

fiftydinar committed:
docs(brew): Include explanation on why manual uninstallation is necessary

Also give big thanks to M2

fiftydinar committed:
docs(brew): Further explain on why the manual uninstallation is needed

fiftydinar committed:
docs(brew): Add link to files module documentation

fiftydinar committed:
docs(brew): Be more detailed about how the module works

fiftydinar committed:
docs(brew): Some small fixes & restore credits

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+# brew
+
+The brew module installs [Homebrew (Brew)](https://brew.sh/) on your system and ensures the package manager remains updated and maintained. This module sets up systemd services to periodically update and upgrade the installed Brew packages.
```
```suggestion
The brew module installs [Homebrew/Linuxbrew](https://brew.sh/) on your system and ensures the package manager remains updated and maintained. This module also sets up systemd services to periodically update the installed Brew packages.
```

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+- Installs Brew at build-time.
+- Sets up systemd services to update the Brew binary to the latest version.
+- Sets up systemd services to upgrade Brew packages automatically.
+- Options to control the frequency of updates and upgrades.
```
```suggestion
## Features
- Installs Brew at build-time.
- Sets up systemd services to automatically update Brew to the latest version.
- Sets up systemd services to automatically upgrade Brew packages.
- Sets up bash and fish completions for Brew.
```

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+
+**Rest of the setup:**
+- `brew-update` is runnning in specified time to update Brew binary to latest version
+- `brew-upgrade` is runnning in specified time to upgrade Brew packages to latest versions
```
```suggestion
- `brew-update` is runs at the specified time to update Brew to the latest version
- `brew-upgrade` is runs at the specified time to upgrade Brew packages```

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+If false, disables automatic activation of `brew-update.timer`.
+
+### `auto-upgrade` (optional: boolean, default: true)
+If false, disables automatic activation of `brew-upgrade.timer`.
```
Move these above the wait timers, they will probably be the most sought after.

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+If false, disables automatic activation of `brew-upgrade.timer`.
+
+!!! warning
+    Please review the brew documentation carefully before modifying these settings.
```
```suggestion
    Please review the Brew documentation carefully before modifying the settings below.
```

fiftydinar committed:
docs(brew): Rename Brew to Linuxbrew

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
docs(brew): Change features wording

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
docs(brew): Reword update & upgrade wording

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+    Please review the brew documentation carefully before modifying these settings.
+
+### `nofile-limits` (optional: boolean, default: false)
+Determines whether to apply nofile limits for Brew installations. When set to true, it increases the nofile limits to prevent certain packages from failing due to file limits. However, it's important to note that increasing nofile limits can have potential security implications. Defaults to false for security purposes.
```
I have no idea what `nofile limits` and what that would mean for me. I'd also be interested in knowing if these limits are set by default when installing brew the official way. I googled the term, but did not find any information about security implications. 

I also assume that the warning above applies to this setting mostly, so a link to the referenced Brew documentation would be great, if that exists.

fiftydinar replied:
It is not set by default in `brew`.

Homebrew doesn't have documentation about this either, nor they are not aware about this afaik.

It is something that Bluefin maintainers noticed, where some packages installed from `brew` complain about reached nofile limits & refuse to work due to this.
https://github.com/ublue-os/bluefin/pull/988

This doesn't just happen with `brew` packages in particular, it can happen with other packages too.
Why this happens in more detail, I have no idea.

About security implications, this is just my own view of it. It is indeed true that there are no searchable topics regarding security of nofile limits.

Maybe @qoijjj knows something more about this.

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+Determines whether to apply nofile limits for Brew installations. When set to true, it increases the nofile limits to prevent certain packages from failing due to file limits. However, it's important to note that increasing nofile limits can have potential security implications. Defaults to false for security purposes.
+
+### `brew-analytics` (optional: boolean, default: true)
+Determines whether to opt-out of Brew analytics. When set to true, analytics are enabled.
```
Is this in scope for the warning above? If not, consider moving it above the warning. IMO analytics opt-out is more user preference than something that should be warned about.

fiftydinar replied:
Yes, I can move this.

fiftydinar committed:
docs(brew): Move update & upgrade service labels above timers

fiftydinar committed:
docs(brew): Change brew documentation wording

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+When excluding `brew` module from the recipe, it's not enough to get it removed.  
+On booted system, it's also necessary to run the `brew` uninstalation script.
+
+Either local-user can execute this script manually or image-maintainer can make it automatic through SystemD service.
```
```suggestion
Removing the `brew` module from the recipe is not enough to get it completely removed.   
On a booted system, it's also necessary to run the `brew` uninstallation script.

Either a local-user can execute this script manually or the image-maintainer may make it automatic through a custom systemd service.
```

xynydev commented on modules/brew/README.md:
````diff
@@ -0,0 +1,136 @@
+Either local-user can execute this script manually or image-maintainer can make it automatic through SystemD service.
+
+Uninstallation script:  
+```
````
```suggestion
```bash
```

(syntax highlighting)

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+- Sets up systemd services to upgrade Brew packages automatically.
+- Options to control the frequency of updates and upgrades.
+
+## How it works
```
All these implementation details are great and I enjoyed reading it. However, is it more important than the available configuration options? I'm not sure how I would solve this problem, though, since I have [an open issue](https://github.com/blue-build/website/issues/55) for generating the configuration options -section from the schema. Maybe it's not urgent right now and I can solve it later...

fiftydinar replied:
Maybe I can move the section down below. I thought about it, but forgot it.

xynydev replied:
I'll mark this as resolved, it's not that important.

fiftydinar committed:
docs(brew): Syntax highlighting for bash script

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
docs(brew): Improve uninstallation intro wording

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+If false, disables automatic activation of `brew-upgrade.timer`.
+
+### `wait-after-boot-upgrade` (optional: string, default: '30min')
+Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).
```
Ok, how about this ordering:
- `auto-update`
   - need to add definition of `update` to distinguish from `upgrade`
- `update-interval`
- `wait-after-boot-update`
   - feels like this could be renamed to `update-wait-after-boot`
- `auto-upgrade`
   - need to add definition of `upgrade` to distinguish from `update`
- `upgrade-interval`
- `wait-after-boot-upgrade`
   - feels like this could be renamed to `upgrade-wait-after-boot`

fiftydinar replied:
I agree with this.

I will change the code to also reflect this.

fiftydinar replied:
About

`need to add definition of upgrade to distinguish from update`

Idk what would be the best suggestion.

fiftydinar committed:
docs(brew): Move brew-analytics above warning

fiftydinar committed:
docs(brew): Update update/upgrade ordering pt. 1

fiftydinar committed:
docs(brew): Update update/upgrade naming pt. 2

fiftydinar committed:
docs(brew): Update update/upgrade order pt. 3

xynydev commented on modules/brew/brew-bash-completions.sh:
```diff
@@ -0,0 +1,26 @@
+# shellcheck shell=sh disable=SC1091,SC2039,SC2166
+
+# Check for interactive bash and that we haven't already been sourced.
+if [ "x${BASH_VERSION-}" != x -a "x${PS1-}" != x -a "x${BREW_BASH_COMPLETION-}" = x ]; then
```
I'd like this shellcheck error fixed https://github.com/koalaman/shellcheck/wiki/SC2268

fiftydinar replied:
I think it should be like this:

```
if [[ "${BASH_VERSION-}" != "" ]] && [[ "${PS1-}" != "" ]] && [[ "${BREW_BASH_COMPLETION-}" == "" ]]; then
```

This bash completion is copied straight from Bluefin.
Not sure if Bluefin also copied it from somewhere, but I know this is not from official brew.
I think it's also custom made.

Idk why it has so many `sh`-isms tbh

fiftydinar replied:
Maybe it's compatibility if someone uses POSIX mode in Bash?

fiftydinar replied:
But it also uses arrays, so it doesn't make sense that it's for `sh`, which doesn't have array functionality.

fiftydinar replied:
I'm changing it back, since `bash-completions` repo uses this mechanism to combat `zsh` compatibility (which can often source from `/etc/profile.d`)

xynydev commented on modules/brew/brew-bash-completions.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
```
```suggestion
#!/usr/bin/env bash
```

switch to bash shebang

this script seems to have multiple bashisms that shellcheck errors on thanks to the wrong shebang

xynydev commented on modules/brew/brew-bash-completions.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+# shellcheck shell=sh disable=SC1091,SC2039,SC2166
```
This line seems to set the shell to `sh` explicitly? @mecattaf is this intentional? Also see my other comment https://github.com/blue-build/modules/pull/239/files#r1622418859

```suggestion
# shellcheck shell=bash disable=SC1091,SC2166	
```

https://github.com/koalaman/shellcheck/wiki/SC2039 not necessary to ignore if using actually bash (which we are, these are bash completions and contain multiple bashisms)

fiftydinar replied:
@xynydev `sh` in Fedora is a symlink to `bash`. So this is effectively `bash`.

But I agree that setting it explicitely to `bash` is better.

xynydev commented on modules/brew/brew-fish-completions.fish:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/fish
```
```suggestion
#!/usr/bin/env fish
```

fiftydinar committed:
chore(brew): Use `#!/usr/bin/env fish` instead of `#!/usr/bin/fish`

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
chore(brew): Use `#!/usr/bin/env bash` instead of `#!/bin/sh`

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
chore(brew): Set shellcheck from sh to bash

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
fix(brew): Convert SH-ism to bash in bash completion

fiftydinar committed:
chore(brew): Use double brackets in bash shell completion

fiftydinar committed:
chore(brew): Make rc in brackets

fiftydinar committed:
chore(brew): Use more double brackets

fiftydinar committed:
chore(brew): Add more double brackets pt. 2

fiftydinar committed:
chore(brew): Make bash shell completion spacing better

fiftydinar committed:
chore(brew): Revert to `bash-completions` completion

fiftydinar committed:
docs(brew): Add comment about why bash completion uses sh

fiftydinar commented on modules/brew/README.md:
```diff
@@ -0,0 +1,136 @@
+# brew
```
I just remembered that `brew.sh` is included in `ublue-os-just` RPM package.
It resides in `/usr/etc/profile.d/brew.sh`

It adds brew in path only when shell is interactive, solving issues with `brew` path clashing with other programs

```bash
#!/usr/bin/env bash
[[ -d /home/linuxbrew/.linuxbrew && $- == *i* ]] && eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
```

fish already contains this part in `brew-fish-completions`.

This is only in Universal Blue images.

We need to account for other images which don't include this.

fiftydinar committed:
fix(brew): Apply brew path clash fix

fiftydinar committed:
docs(brew): Explain brew path clash fix

fiftydinar committed:
docs(brew): Add file location of brew path fix

fiftydinar committed:
docs(brew): Reword exporting brew path

fiftydinar committed:
docs(brew): Reword again

fiftydinar committed:
docs(brew): Reword...

fiftydinar committed:
chore(brew): Copy bash completion to `/usr/etc` instead of `/etc`

fiftydinar committed:
chore(brew): Make `/usr/etc/profile.d` if it doesn't exist

fiftydinar committed:
docs(brew): Fix typo in README

fiftydinar committed:
docs(brew): Use caution card instead of warning string

xynydev commented on modules/brew/README.md:
```diff
@@ -22,6 +22,8 @@ The brew module installs [Homebrew/Linuxbrew](https://brew.sh/) on your system a
 - Brew in `/home/linuxbrew/` is compressed in tar, copied to `/usr/share/homebrew/` & permissions to it are set to default user (UID 1000)
 - `brew-update` & `brew-upgrade` SystemD service timers are enabled (by default)
+- Applied fix for program path conflict between system & brew packages with same name  
+  by exporting Brew path to shell environment when interactive shell is used only, unlike offical Homebrew way of exporting it globally (`/etc/profile.d/brew.sh`)
```
```suggestion
- A fix for path conflicts between system & brew packages with the same name is applied by adding Brew to path only in interactive shells, unlike what Brew does by default.
```

xynydev commented on modules/brew/brew.sh:
```diff
@@ -179,10 +179,23 @@ Persistent=true
 
+# Apply brew shell environment only when shell is interactive
+# Fish already includes this fix in brew-fish-completions.sh
+# Officially, brew applies brew shell environment globally, which causes path conflicts between system & brew installed programs with same name.
```
```suggestion
# By default Brew applies the shell environment changes globally, which causes path conflicts between system & brew installed programs with same name.
```

xynydev commented:
This is starting to look pretty good. I'd like the `nofile-limits` explanation be expanded before merge, though. One of y'all would probably qualify better to do that.

fiftydinar committed:
docs(brew): Brew path reword

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
chore(brew): Implement code comment fix

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
docs(brew): Try to improve `nofile-limits` section

fiftydinar commented:
@xynydev Can you check the documentation now for `nofile-limits`?

It maybe can be better reworded, but I think it explains things a bit better.

fiftydinar committed:
docs(brew): Use serverfault link instead

fiftydinar committed:
docs(brew): Thread, not article

fiftydinar committed:
Merge branch 'main' into add-brew-module

xynydev commented:
It looks great! I'll have one final read of the docs before approving.

xynydev commented on modules/brew/README.md:
```diff
@@ -0,0 +1,139 @@
+Defines how often the Brew upgrade service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).
+
+### `upgrade-wait-after-boot` (optional: string, default: '30min')
+Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).
```
Somewhere around here the difference between updates and upgrades should be explained.

fiftydinar replied:
I got the better idea. Please check out this commit:

[`8ad2c3e` (#239)](https://github.com/blue-build/modules/pull/239/commits/8ad2c3ec05e2b401466186e36c101e9051c811dd)

fiftydinar committed:
docs(brew): Improve Configuration section

fiftydinar committed:
docs(brew): Than vs then

Love to mess this one

fiftydinar committed:
docs(brew): Uninstall script wording fix

xynydev commented:
@fiftydinar since you've made change reqs we need your manual approval to merge.

xynydev commented:
Ah, @gmpinder requested changes too. I don't think he's awake yet so I guess I'll just force-merge.
